ethcore/src/blockchain/blockchain.rs

```diff
@@ -1005,29 +1006,50 @@ impl BlockChain {
 			batch.extend_with_cache(db::COL_EXTRA, &mut *write_receipts, update.block_receipts, CacheUpdatePolicy::Remove);
 		}
 
-		{
-			let mut write_blocks_blooms = self.blocks_blooms.write();
-			batch.extend_with_cache(db::COL_EXTRA, &mut *write_blocks_blooms, update.blocks_blooms, CacheUpdatePolicy::Remove);
-		}
-
 		// These cached values must be updated last with all four locks taken to avoid
 		// cache decoherence
 		{
 			let mut best_block = self.pending_best_block.write();
+			let mut write_blocks_blooms = self.blocks_blooms.write();
 			// update best block
 			match update.info.location {
 				BlockLocation::Branch => (),
-				_ => if is_best {
-					batch.put(db::COL_EXTRA, b"best", &update.info.hash);
-					*best_block = Some(BestBlock {
-						hash: update.info.hash,
-						number: update.info.number,
-						total_difficulty: update.info.total_difficulty,
-						timestamp: update.timestamp,
-						block: update.block.to_vec(),
-					});
+				BlockLocation::BranchBecomingCanonChain(_) => {
+					// clear all existing blooms, cause they may be created for block
+					// number higher than current best block
+					*write_blocks_blooms = update.blocks_blooms;
```

Comment:
So we override `write_blocks_blooms`, but we don't really clear anything from the `batch`. So we may end up writing some additional data to the database, right? I guess it's not harmful?

Reply:
On a second thought:

Why do we actually clear everything from `write_blocks_blooms`? It's actually a cache of whatever is in the database, no?

So do I understand correctly that previously:
1. The cache was reflecting what was in the db (since we always did `extend_with_cache`)
2. The problem was that some keys were overwritten instead of `accrued` (that's what the `CanonChain` branch is doing)
3. Currently we may end up clearing the cache, but at least the database should contain correct values?

Could you also elaborate on what is actually stored in `update.blocks_blooms` (what is used as a `key` and `value`)?

Reply:
> So we override `write_blocks_blooms`, but we don't really clear anything from the `batch`. So we may end up writing some additional data to the database, right? I guess it's not harmful?

There is nothing in the batch to be cleared.

> Why do we actually clear everything from write_blocks_blooms? It's actually a cache of whatever is in the database, no?

cause new `BlockLocation` is `BranchBecomingCanonChain`. Database entries need to be updated and so is cache.

> 1. The cache was reflecting what was in the db (since we always did extend_with_cache)

No, it was always empty after executing this function, cause `UpdatePolicy` was `Remove`.

https://github.com/paritytech/parity/blob/5b4abec2dbc3f5653b2caf2cc481411a29c7d00b/ethcore/src/db.rs#L127-L131

2. The problem was that some keys were overwritten instead of `accrued` (that's what the `CanonChain` branch is doing)

No. The problem was that there were never in cache and not yet in database, cause noone has written transaction batch to it.

3. Currently we may end up clearing the cache, but at least the database should contain correct values?

We do not care about database, cause everything is in the cache (as it should be).